### PR TITLE
Add retry for ctr snapshot view

### PR DESCRIPTION
**Pull request by ruiwen-zhao**
Sometimes the `ctr snapshot view` command will silently fail (succeed without creating the snapshot). Adding retry around the view logic

See the following log for the repro and how this change fixed it. Look for "Failed to get mount point " and "retry"

```
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: ctr: snapshot tmp_sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 does not exist: not found
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: Failed to get mount point for tmp_sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691. Will retry. 5 retries left.
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: All snapshots:
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: KEY                                                                     PARENT                                                                  KIND
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 sha256:9969d65d0618198425e0be29ec6194be71b36c856ce9fe2142dd54163aef6eae Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:32b31b2b0563a28eb36f90bcd835540ead275b63cf48ed85e998a8df0b1845ff sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:32cf0b3c70998330104acfc1a02f6aae2f8a728bc6ad91c2415fda501593fa81 sha256:79cae367d5d42d54224c249db6a06234d077ea32168bed4261141c1016469623 Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:79cae367d5d42d54224c249db6a06234d077ea32168bed4261141c1016469623 sha256:32b31b2b0563a28eb36f90bcd835540ead275b63cf48ed85e998a8df0b1845ff Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:9969d65d0618198425e0be29ec6194be71b36c856ce9fe2142dd54163aef6eae sha256:e9c1515f88f6e5652bde5583f0d1ddb7dd4097380aa66d90b722c53fbea3ebab Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:e9c1515f88f6e5652bde5583f0d1ddb7dd4097380aa66d90b722c53fbea3ebab sha256:ec983b16636050e69677eb81537e955ab927757c23aaf73971ecf5f71fcc262a Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: sha256:ec983b16636050e69677eb81537e955ab927757c23aaf73971ecf5f71fcc262a                                                                         Committed
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: ctr: failed to remove "tmp_sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691": snapshot tmp_sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 does not exist: not found
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 snapshots/4/fs to Metadata file
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:32b31b2b0563a28eb36f90bcd835540ead275b63cf48ed85e998a8df0b1845ff
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:32b31b2b0563a28eb36f90bcd835540ead275b63cf48ed85e998a8df0b1845ff snapshots/5/fs to Metadata file
Nov 15 01:31:49 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:32cf0b3c70998330104acfc1a02f6aae2f8a728bc6ad91c2415fda501593fa81
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:32cf0b3c70998330104acfc1a02f6aae2f8a728bc6ad91c2415fda501593fa81 snapshots/7/fs to Metadata file
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:79cae367d5d42d54224c249db6a06234d077ea32168bed4261141c1016469623
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:79cae367d5d42d54224c249db6a06234d077ea32168bed4261141c1016469623 snapshots/6/fs to Metadata file
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:9969d65d0618198425e0be29ec6194be71b36c856ce9fe2142dd54163aef6eae
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:9969d65d0618198425e0be29ec6194be71b36c856ce9fe2142dd54163aef6eae snapshots/3/fs to Metadata file
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:e9c1515f88f6e5652bde5583f0d1ddb7dd4097380aa66d90b722c53fbea3ebab
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:e9c1515f88f6e5652bde5583f0d1ddb7dd4097380aa66d90b722c53fbea3ebab snapshots/2/fs to Metadata file
Nov 15 01:31:50 debian google_metadata_script_runner[833]: startup-script-url: Processing sha256:ec983b16636050e69677eb81537e955ab927757c23aaf73971ecf5f71fcc262a
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: Appending sha256:ec983b16636050e69677eb81537e955ab927757c23aaf73971ecf5f71fcc262a snapshots/1/fs to Metadata file
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: Processing is done.
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: Removing the original pulled image docker.io/library/nginx:latest ...
Nov 15 01:31:51 debian containerd[3419]: time="2023-11-15T01:31:51.122512747Z" level=info msg="ImageDelete event &ImageDelete{Name:docker.io/library/nginx:latest,XXX_unrecognized:[],}"
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: docker.io/library/nginx:latest
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: Content of snapshots.metadata file:
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:203226d708244adf6c392a28e22e3a7e276270da43094205137c404fc9949691 snapshots/4/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:32b31b2b0563a28eb36f90bcd835540ead275b63cf48ed85e998a8df0b1845ff snapshots/5/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:32cf0b3c70998330104acfc1a02f6aae2f8a728bc6ad91c2415fda501593fa81 snapshots/7/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:79cae367d5d42d54224c249db6a06234d077ea32168bed4261141c1016469623 snapshots/6/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:9969d65d0618198425e0be29ec6194be71b36c856ce9fe2142dd54163aef6eae snapshots/3/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:e9c1515f88f6e5652bde5583f0d1ddb7dd4097380aa66d90b722c53fbea3ebab snapshots/2/fs
Nov 15 01:31:51 debian google_metadata_script_runner[833]: startup-script-url: sha256:ec983b16636050e69677eb81537e955ab927757c23aaf73971ecf5f71fcc262a snapshots/1/fs
```

### CODE REVIEW

File: gke-disk-image-builder/script/startup.sh [hunk 1]
```diff
@@ -84,14 +84,32 @@ function process_snapshots() {
...
-    # Check if the view was successfully created
-    if [ $? -ne 0 ]; then
-      echo Failed to create snapshot view for $snapshot. Please rerun the tool to try it again.
+    # Add retry in case `ctr snapshot view` fails silently
+    local retries=5
+    while [ ${retries} -ge 0 ]; do 
+      sudo ctr -n k8s.io snapshot view tmp_$snapshot $snapshot
+      # Check if the view was successfully created
+      if [ $? -ne 0 ]; then
+        echo Failed to create snapshot view for $snapshot. Will retry.
```
Comment (elfinhe) [inline gke-disk-image-builder/script/startup.sh:93]: - Please also echo the attempt#.
- Also maybe sleep for 1 second.

File: gke-disk-image-builder/script/startup.sh [hunk 2]
```diff
@@ -84,14 +84,32 @@ function process_snapshots() {
...
+      
+      original_path=$(sudo ctr -n k8s.io snapshot mount /tmp_$snapshot tmp_$snapshot | grep -oP '/\S+/snapshots/[0-9]+/fs' | tr ':' '\n' | head -n 1)
+      if [[ -n "$original_path" ]]; then
+        break
+      fi
+      echo "Failed to get mount point for tmp_$snapshot. Will retry. ${retries} retries left."
+      echo "All snapshots:"
+      sudo ctr -n k8s.io snapshot list
+      sudo ctr -n k8s.io snapshot rm tmp_$snapshot
+      ((retries--))
```
Comment (elfinhe) [inline gke-disk-image-builder/script/startup.sh:105]: Maybe move `((retries--))` to the beginning of the loop. The `continue` branch ("Failed create snapshot view") does not change loop counter $retries, and that could cause endless loop I guess.

Comment (ruiwen-zhao) [reply gke-disk-image-builder/script/startup.sh:105]: uh good catch. Done

File: gke-disk-image-builder/script/startup.sh (hunk 1 above)
Comment (ruiwen-zhao) [reply gke-disk-image-builder/script/startup.sh:93]: done

File: gke-disk-image-builder/script/startup.sh [hunk 3]
```diff
@@ -84,14 +84,33 @@ function process_snapshots() {
...
+      
+      original_path=$(sudo ctr -n k8s.io snapshot mount /tmp_$snapshot tmp_$snapshot | grep -oP '/\S+/snapshots/[0-9]+/fs' | tr ':' '\n' | head -n 1)
+      if [[ -n "$original_path" ]]; then
+        break
+      fi
+      echo "Failed to get mount point for tmp_$snapshot. Will retry. ${retries} retries left."
+      echo "All snapshots:"
+      sudo ctr -n k8s.io snapshot list
+      sudo ctr -n k8s.io snapshot rm tmp_$snapshot
+      sleep 1
```
Comment (elfinhe) [inline gke-disk-image-builder/script/startup.sh:106]: Move the sleep to the continue branch? Other parts look good!

**Review by elfinhe — APPROVED** [review]

File: gke-disk-image-builder/script/startup.sh [hunk 4]
```diff
@@ -84,14 +84,33 @@ function process_snapshots() {
...
+    done
+
+    if [ ${retries} -lt 0 ]; then
+      echo Failed to get snapshot directory for snapshot $snapshot. Please rerun the tool to try it again.
```
Comment (samuelkarp) [inline gke-disk-image-builder/script/startup.sh:110]: With `((retries--))` at the top of the loop, the last iteration will leave the `retries` as `-1`, regardless of whether or not it succeeded.

```
$ cat retries.sh 
#!/bin/bash

retries=5
while [ ${retries} -ge 0 ]; do
  ((retries--))
  if [ ${retries} -ge 0 ]; then
    echo "Trying again, ${retries} left"
    continue
  fi
  echo "Last retry, success"
done

if [ ${retries} -lt 0 ]; then
  echo "Oops, all retries exhausted"
fi 
$ ./retries.sh 
Trying again, 4 left
Trying again, 3 left
Trying again, 2 left
Trying again, 1 left
Trying again, 0 left
Last retry, success
Oops, all retries exhausted
```

File: gke-disk-image-builder/script/startup.sh (hunk 3 above)
Comment (ruiwen-zhao) [reply gke-disk-image-builder/script/startup.sh:106]: I dont think the continue branch needs the sleep. The continue happens when `ctr view` returns an error, in which case the active snapshot will definitely not be created, and we dont need to sleep

File: gke-disk-image-builder/script/startup.sh (hunk 4 above)
Comment (ruiwen-zhao) [reply gke-disk-image-builder/script/startup.sh]: uh nice catch. I changed the entry condition of the while loop to >= 1. And changed the if-condition at the end to check `original_path`

File: gke-disk-image-builder/script/startup.sh [hunk 5]
```diff
@@ -84,14 +84,33 @@ function process_snapshots() {
   for snapshot in "${snapshots[@]}"; do
     echo Processing $snapshot
 
-    sudo ctr -n k8s.io snapshot view tmp_$snapshot $snapshot
-    # Check if the view was successfully created
-    if [ $? -ne 0 ]; then
-      echo Failed to create snapshot view for $snapshot. Please rerun the tool to try it again.
+    # Add retry in case `ctr snapshot view` fails silently
+    local retries=5
+    while [ ${retries} -ge 1 ]; do
```
Comment (samuelkarp) [inline gke-disk-image-builder/script/startup.sh:89]: nit: I'd probably do `-gt 0` but both are effectively the same.

**Review by samuelkarp — APPROVED** [review]

**Pull request merged**